### ISSUE TITLE
In-depth review of the locks (we have deadlocks/unlocking of unlocked mutex)

### ISSUE THREAD

**Issue by lbarman** [lbarman/prifi#46]

**Comment by lbarman** [comment]
This is completely broken. Stuff like : 
```
...
// changing state
p.relayState.locks.state.Lock()
p.relayState.currentState = RELAY_STATE_COLLECTING_SHUFFLES	p.relayState.locks.state.Unlock()
...
```
does *not* work, as in the rest of the function, currentState is unvalid / undefined w.r.t concurrent changes. At least, we need to lock everything at the beginning of the function.

**Comment by lbarman** [comment]
- [ ]

**Comment by lbarman** [comment]
Solved the hard way 2e500e3648d266de4bd91b39b9dae585572e6723
